exporter/loadbalancingexporter/exporter.go

```diff
@@ -16,42 +16,208 @@ package loadbalancingexporter
 
 import (
 	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
 
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
 	"go.opentelemetry.io/collector/component"
 	"go.opentelemetry.io/collector/config/configmodels"
 	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/exporter/otlpexporter"
 	"go.uber.org/zap"
 )
 
 var _ component.TracesExporter = (*exporterImp)(nil)
 
+const (
+	defaultPort = "55680"
+)
+
+var errNoResolver = errors.New("no resolvers specified for the exporter")
+
 type exporterImp struct {
 	logger *zap.Logger
 	config Config
+
+	res  resolver
+	ring *hashRing
+
+	exporters            map[string]component.TracesExporter
+	exporterFactory      component.ExporterFactory
+	templateCreateParams component.ExporterCreateParams
+
+	stopped    bool
+	shutdownWg sync.WaitGroup
+	updateLock sync.RWMutex
 }
 
 // Crete new exporter
 func newExporter(params component.ExporterCreateParams, cfg configmodels.Exporter) (*exporterImp, error) {
 	oCfg := cfg.(*Config)
 
+	tmplParams := component.ExporterCreateParams{
+		Logger:               params.Logger,
+		ApplicationStartInfo: params.ApplicationStartInfo,
+	}
+
+	var res resolver
+	if oCfg.Resolver.Static != nil {
+		var err error
+		res, err = newStaticResolver(oCfg.Resolver.Static.Hostnames)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if oCfg.Resolver.DNS != nil { // if both are configured, DNS takes precedence
+		dnsLogger := params.Logger.With(zap.String("resolver", "dns"))
+
+		var err error
+		res, err = newDNSResolver(dnsLogger, oCfg.Resolver.DNS.Hostname)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if res == nil {
+		return nil, errNoResolver
+	}
+
 	return &exporterImp{
 		logger: params.Logger,
 		config: *oCfg,
+
+		res: res,
+
+		exporters:            map[string]component.TracesExporter{},
+		exporterFactory:      otlpexporter.NewFactory(),
+		templateCreateParams: tmplParams,
 	}, nil
 }
 
 func (e *exporterImp) Start(ctx context.Context, host component.Host) error {
+	e.res.onChange(e.onBackendChanges)
+	if err := e.res.start(ctx); err != nil {
+		return err
+	}
+
 	return nil
 }
 
+func (e *exporterImp) onBackendChanges(resolved []string) {
+	resolved = sort.StringSlice(resolved)
+	newRing := newHashRing(resolved)
+
+	if !newRing.equal(e.ring) {
+		e.updateLock.Lock()
+		defer e.updateLock.Unlock()
+
+		e.ring = newRing
+
+		// TODO: set a timeout?
+		ctx := context.Background()
+
+		// add the missing exporters first
+		e.addMissingExporters(ctx, resolved)
```

Comment:
If I understand correctly, atomicity is important. Should we prepare an `exporters` local variable and then set `e.exporters` in one call so there's no in-progress state?

Reply:
The ring itself has to be atomic, but the list of exporters doesn't have to be that strict. The alternative that I can think of would make the list of exporters duplicated: one as a parameter to the `e.addMissingExporters`, and one as part of the `e` exporter itself. As this is not in the hot path, the cost of holding the lock for a bit longer is justifiable, IMO.